File: backend/src/repository/user_repository.rs
use async_trait::async_trait;
use uuid::Uuid;

use crate::domain::entity::user::User;
use crate::domain::valueobject::email::Email;
use crate::domain::valueobject::username::Username;

/// ユーザーリポジトリのエラー型
#[derive(Debug, thiserror::Error)]
pub enum UserRepositoryError {
    #[error("User not found")]
    NotFound,
    #[error("User already exists")]
    AlreadyExists,
    #[error("Database error: {0}")]
    DatabaseError(String),
}

/// ユーザーリポジトリのトレイト
#[async_trait]
pub trait UserRepository: Send + Sync {
    /// ユーザーを保存
    async fn save(&self, user: &User) -> Result<(), UserRepositoryError>;

    /// IDでユーザーを取得
    async fn find_by_id(&self, user_id: &Uuid) -> Result<Option<User>, UserRepositoryError>;

    /// メールアドレスでユーザーを取得
    async fn find_by_email(&self, email: &Email) -> Result<Option<User>, UserRepositoryError>;

    /// ユーザー名でユーザーを取得
    async fn find_by_username(
        &self,
        username: &Username,
    ) -> Result<Option<User>, UserRepositoryError>;

    /// メールアドレスが既に登録されているか確認
    async fn exists_by_email(&self, email: &Email) -> Result<bool, UserRepositoryError>;

    /// ユーザー名が既に登録されているか確認
    async fn exists_by_username(&self, username: &Username) -> Result<bool, UserRepositoryError>;
}

File: backend-hono/src/repository/interface/UserRepository.ts
import { User } from '../../domain/entity/User';

export class UserRepositoryError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'UserRepositoryError';
  }
}

export interface UserRepository {
  save(user: User): Promise<void>;
  findById(userId: string): Promise<User | null>;
  findByEmail(email: string): Promise<User | null>;
  findByUsername(username: string): Promise<User | null>;
  existsByEmail(email: string): Promise<boolean>;
  existsByUsername(username: string): Promise<boolean>;
}
